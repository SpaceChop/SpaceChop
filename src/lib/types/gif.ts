import ImageDefinition from '../../imagedef';

/**
 * @see http://www.onicos.com/staff/iz/formats/gif.html
 * @see http://giflib.sourceforge.net/whatsinagif/bits_and_bytes.html
 *
 * GIF format
 * Byte Order: Little-endian
 * GIF Header
 * Offset   Length   Contents
 *   0      3 bytes  "GIF" 0x47 0x49 0x46
 *   3      3 bytes  "87a" or "89a"  0x38 0x39|0x37 0x61
 *   6      2 bytes  <Logical Screen Width>
 *   8      2 bytes  <Logical Screen Height>
 *  10      1 byte   bit 0:    Global Color Table Flag (GCTF)
 *                   bit 1..3: Color Resolution
 *                   bit 4:    Sort Flag to Global Color Table
 *                   bit 5..7: Size of Global Color Table: 2^(1+n)
 *  11      1 byte   <Background Color Index>
 *  12      1 byte   <Pixel Aspect Ratio>
 *  13      ? bytes  <Global Color Table(0..255 x 3 bytes) if GCTF is one>
 *          ? bytes  <Blocks>
 *          1 bytes  <Trailer> (0x3b)
 * *
 */
export default (buffer): ImageDefinition => {
  const BLOCK_TERMINATOR = 0x00;
  const EXTENSION_INTRODUCER = 0x21;
  const GRAPHIC_CONTROL_LABEL = 0xf9;
  const TRANSPARENT_COLOR_FLAG = 0x01;
  const IMAGE_SEPARATOR = 0x2c;
  const INTERLACE_FLAG = 0x64;
  if (buffer.length < 13 ||
      // "GIF" 0x47 0x49 0x46
      buffer[0] !== 0x47 || buffer[1] !== 0x49 || buffer[2] !== 0x46 ||
      // "87a" or "89a"
      buffer[3] !== 0x38 || (buffer[4] !== 0x39 && buffer[4] !== 0x37) || buffer[5] !== 0x61) {
    return;
  }
  const width = buffer.readUInt16LE(6);
  const height = buffer.readUInt16LE(8);
  let interlacing = false;
  let alpha = false;
  let animated = false;

  // version: 8[9]a
  if (buffer[4] === 0x39) {
    // graphic control extension.
    let frames = 0;
    for (let i = 14; i < buffer.length; i++) {
      // BLOCK_TERMINATOR
      if (buffer[i] === BLOCK_TERMINATOR &&
          buffer[i + 1] === EXTENSION_INTRODUCER &&
          buffer[i + 2] === GRAPHIC_CONTROL_LABEL) {
        frames++;
        alpha = !!(buffer[i + 4] & TRANSPARENT_COLOR_FLAG); // tslint:disable-line
        if (frames > 1) {
          animated = true;
          break;
        }
      }
      if (buffer[i] === BLOCK_TERMINATOR && buffer[i + 1] === IMAGE_SEPARATOR) {
        interlacing = !!(buffer[i + 10] & INTERLACE_FLAG); // tslint:disable-line
      }
    }
  }

  return {
    width,
    height,
    type: 'gif',
    alpha,
    animated,
    interlacing,
    size: buffer.length,
  };
};
