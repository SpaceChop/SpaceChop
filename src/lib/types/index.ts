import { Stream } from 'stream';
import ImageDefinition from '../../imagedef';
import isGIF from './gif';
import isJPEG from './jpeg';
import isPNG from './png';
import isWEBP from './webp';

const tests = [
  isGIF,
  isJPEG,
  isPNG,
  isWEBP,
];

export default (stream: Stream): Promise<any> => new Promise((resolve) => {
  const chunks = [];
  stream.on('data', (chunk) => {
    chunks.push(chunk);
  });
  stream.on('end', async () => {
    const buffer = Buffer.concat(chunks);
    for (const test of tests) {
      const result: ImageDefinition = test(buffer);
      if (result) {
        resolve(result);
        break;
      }
    }
  });
});
