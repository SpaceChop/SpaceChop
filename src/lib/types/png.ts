import ImageDefinition from '../../imagedef';

/**
 * @see http://en.wikipedia.org/wiki/Portable_Network_Graphics
 *
 * A PNG file starts with an 8-byte signature. The hexadecimal byte values are
 * 89 50 4E 47 0D 0A 1A 0A; the decimal values are 137 80 78 71 13 10 26 10.
 * Each of the header bytes is there for a specific reason:[7]
 * Bytes Purpose
 * 89  Has the high bit set to detect transmission systems that do not support 8
 * bit data and to reduce the chance that a text file is mistakenly interpreted
 * as a PNG, or vice versa.
 * 50 4E 47  In ASCII, the letters PNG, allowing a person to identify the format
 * easily if it is viewed in a text editor.
 * 0D 0A A DOS-style line ending (CRLF) to detect DOS-Unix line ending conversion of the data.
 * 1A  A byte that stops display of the file under DOS when the command type has been usedâ€”the end-of-file character
 * 0A  A Unix-style line ending (LF) to detect Unix-DOS line ending conversion.
 */
export default (buffer): ImageDefinition => {
  if (buffer.length < 16 ||
      buffer[0] !== 0x89 ||
      // PNG
      buffer[1] !== 0x50 || buffer[2] !== 0x4E || buffer[3] !== 0x47 ||
      // \r\n
      buffer[4] !== 0x0D || buffer[5] !== 0x0A ||
      buffer[6] !== 0x1A || buffer[7] !== 0x0A) {
    return;
  }
  // Length  Chunk type  Chunk data  CRC
  // 4 bytes 4 bytes Length bytes  4 bytes
  const length = buffer.readUInt32BE(8);
  // var chunkType = buf.slice(12, 16).toString(); // should be 'IHDR'
  const chunkData = buffer.slice(16, 16 + length);
  // Width:              4 bytes   0
  // Height:             4 bytes   4
  // Bit depth:          1 byte    8
  // Color type:         1 byte    9
  // Compression method: 1 byte    10
  // Filter method:      1 byte    11
  // Interlace method:   1 byte    12
  const width = chunkData.readUInt32BE(0, true);
  const height = chunkData.readUInt32BE(4, true);
  const color = chunkData.readUInt8(9, true);
  const interlace = chunkData.readUInt8(12, true);
  const part = buffer.slice(0, 48).toString('utf8');
  return {
    width,
    height,
    type: 'png',
    animated: /acTL/.test(part),
    alpha: !!(color & 4), // tslint:disable-line
    interlacing: interlace > 0,
    size: buffer.length,
  };
};
