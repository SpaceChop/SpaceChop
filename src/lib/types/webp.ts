import ImageDefinition from '../../imagedef';

/**
 * @see https://developers.google.com/speed/webp/docs/riff_container
 *
 * WebP File Header
 *  0                   1                   2                   3
 *  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |      'R'      |      'I'      |      'F'      |      'F'      |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |                           File Size                           |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 * |      'W'      |      'E'      |      'B'      |      'P'      |
 * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
 *
 * 'RIFF': 32 bits
 *   The ASCII characters 'R' 'I' 'F' 'F'.
 * File Size: 32 bits (uint32)
 *   The size of the file in bytes starting at offset 8.
 *   The maximum value of this field is 2^32 minus 10 bytes and
 *     thus the size of the whole file is at most 4GiB minus 2 bytes.
 * 'WEBP': 32 bits
 *   The ASCII characters 'W' 'E' 'B' 'P'.
 */
// const ANMF = new Uint8Array([
//   0x41, 0x4e, 0x4d, 0x46,
// ]);
export default (buffer): ImageDefinition => {
  if (buffer.length < 12 ||
    buffer.slice(0, 4).toString() !== 'RIFF' ||
    buffer.slice(8, 12).toString() !== 'WEBP') {
    return;
  }

  // uint32: A 32-bit, little-endian, unsigned integer.
  // The file size in the header is the total size of the chunks that follow plus 4 bytes for the 'WEBP' FourCC.
  const size = buffer.readUInt32LE(4);
  let lossy = false;
  let animated = false;
  let width = null;
  let height = null;
  let alpha = false;

  let offset = 12;
  while (offset + 8 < buffer.length) {
    const chunkType = buffer.slice(offset, offset + 4).toString();
    const chunkSize = buffer.readUInt32LE(offset + 4);
    switch (chunkType) {
      case 'VP8 ': {
        // 3 bytes - "Uncompressed Data Chunk"
        // 3 bytes - VP8 intraframe start code
        const horizontalSizeCode = buffer.readUInt16LE(offset + 8 + 3 + 3);
        width = horizontalSizeCode & 0x3fff; // tslint:disable-line
        const verticalSizeCode = buffer.readUInt16LE(offset + 8 + 3 + 5);
        height = verticalSizeCode & 0x3fff; // tslint:disable-line
        break;
      }
      case 'VP8L': {
        // 1 byte - One byte signature 0x2f.
        lossy = true;
        const bits = buffer.readUInt32LE(offset + 8 + 1);
        width = (bits & 0x3FFF) + 1; // tslint:disable-line
        height = ((bits >> 14) & 0x3fff) + 1; // tslint:disable-line
        alpha = !!((bits >> 28) & 0x01); // tslint:disable-line
        break;
      }
      case 'VP8X': {
        // 80 bits
        //
        // 2 bits - Reserved (1)
        // 1 bit - ICC Profile flag
        // 1 bit - Alpha flag
        // 1 bit - EXIF metadata flag
        // 1 bit - XMP metadata flag
        // 1 bit - Animation flag
        // 1 bit - Reserved (2)
        // 24 bits - Reserved (3)
        // 24 bits - Canvas Width (actual canvas width should add 1 to read value)
        // 24 bits - Canvas Height (actual canvas height should add 1 to read value)
        const bits = buffer.readUInt8(offset + 8);
        alpha = !!((bits >> 4) & 0x01); // tslint:disable-line
        animated = !!((bits >> 1) & 0x01); // tslint:disable-line
        width = buffer.readUInt16LE(offset + 8 + 4) + 1;
        height = buffer.readUInt16LE(offset + 8 + 7) + 1;
        break;
      }
      // case 'ANIM':
      //   break;
      case 'ANMF':
        animated = true;
        break;
      case 'ALPH':
        alpha = true;
        break;
      // case 'ICCP':
      //   break;
      // case 'EXIF':
      //   break;
      // case 'XMP':
      //   break;
    }
    offset += chunkSize + 8;
  }

  return {
    width,
    height,
    type: 'webp',
    alpha,
    animated,
    interlacing: false,
    lossy,
    size: size + 8,
  };
};
