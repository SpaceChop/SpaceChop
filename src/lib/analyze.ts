import { Stream } from 'stream';
import { DefinitionRequirement, ImageDefinition } from '../types';
import facedetect from './face-detection/detect';
import StreamSwitch from './stream-switch';
import types from './types';

export default async (
  stream: Stream,
  requirements: { [key: number]: DefinitionRequirement },
): Promise<ImageDefinition> => {
  const streamSwitch = new StreamSwitch(stream);
  const streamToAnalyzeTypes = streamSwitch.createReadStream();
  const info = await types(streamToAnalyzeTypes);
  // in case of face detection, analyze image for faces
  for (const i in requirements) {
    if (requirements[i] === 'faces') {
      const streamToAnalyzeFaces = streamSwitch.createReadStream();
      const faces = await facedetect(streamToAnalyzeFaces, info);
      if (faces.length > 0) {
        info.faces = faces;
      }
    }
  }
  return {
    ...info,
  };
};
