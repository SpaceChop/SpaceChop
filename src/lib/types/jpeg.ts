import { ImageDefinition } from '../../types';

export default (buffer): ImageDefinition => {
  if (buffer.length < 3 ||
    buffer[0] !== 255 ||
    buffer[1] !== 216 ||
    buffer[2] !== 255
  ) {
    return;
  }

  let width = null;
  let height = null;
  let interlacing = false;
  let offset = 4;
  while (offset < buffer.length) {
    const i = buffer.readUInt16BE(offset);
    // Every JPEG block must begin with a 0xFF
    if (buffer[offset + i] === 0xFF) {
      const next = buffer[offset + i + 1];
      if (next === 0xC2) {
        interlacing = true;
      }
      if (next === 0xC0 || next === 0xC1 || next === 0xC2) {
        const j = offset + i + 2 + 2 + 1;
        width = buffer.readUInt16BE(j + 2, true);
        height = buffer.readUInt16BE(j, true);
        break;
      }
    }
    // move to the next block
    offset += i + 2;
  }

  return {
    width,
    height,
    type: 'jpeg',
    alpha: false,
    animated: false,
    interlacing,
    size: buffer.length,
  };
};
