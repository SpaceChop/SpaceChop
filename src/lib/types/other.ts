import { Stream } from 'stream';
import { ImageDefinition } from '../../types';
import { spawn } from 'duplex-child-process';

const identify = (input: Buffer): Promise<{
  type: string,
  width: number,
  height: number,
  color: string,
  alpha: boolean,
}[]> => new Promise((resolve, reject) => {
  const command = `identify -`;
  const pattern = /^([^ ]+)+ (\d+)x(\d+) [^ ]+ [^ ]+ (\w+)/i;
  const process = spawn('sh', ['-c', command]);
  process.write(input);
  process.end();
  const chunks = [];
  process.on('data', (chunk) => {
    chunks.push(chunk);
  });
  process.on('error', (err) => reject(err));
  process.on('end', () => {
    const buffer = Buffer.concat(chunks).toString('utf8');
    const images = buffer
      .replace(/-=>[^ ]+/ig, '')
      .trim()
      .split(/\n/)
      .map(image => {
        const [, type, width, height, color] = image.match(pattern);
        const alpha = /a$/.test(color);
        return {
          type: type.toLowerCase(),
          width: parseInt(width),
          height: parseInt(height),
          color,
          alpha,
        };
      });
    resolve(images);
  });
});

export default async (buffer: Buffer): Promise<ImageDefinition> => {
  if (!buffer || buffer.length === 0) {
    return;
  }
  try {
    const images = await identify(buffer);
    const type = images[0].type;
    const width = images[0].width;
    const height = images[0].height;
    const alpha = images.some(image => image.alpha);
    const animated = images.length > 1;
    const size = buffer.length;
    const interlacing = false;

    // Change this to add support for more types.
    if (type === 'heic') {
      return {
        width,
        height,
        type,
        alpha,
        animated,
        interlacing,
        size,
      };
    }
  } catch (err) {
    throw new Error('Unsupported type');
  }
};
