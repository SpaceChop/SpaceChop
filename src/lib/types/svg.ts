import isBinaryFile from 'isbinaryfile';
import XML from 'pixl-xml';
import { ImageDefinition } from '../../types';

export default (buffer): ImageDefinition => {
  
  const bufferByteLength = buffer.byteLength; // Get the number of bytes

  // Check if the data indicates an empty or binary file.
  if (buffer.length < 3 || isBinaryFile.sync(buffer, bufferByteLength)) {
    return;
  }

  let width;
  let height;
  let interlacing = false;
  let alpha = false;
  let animated = false;

  try {
    const svgDoc = XML.parse(buffer.toString());
    let svgHeight;
    let svgWidth;

    // Logic borrowed from https://github.com/fixiecoder/node-svg-dimensions/blob/master/svg-dimensions.js
    const hasWidthHeightAttr = svgDoc.height && svgDoc.width;
    if (hasWidthHeightAttr) {
      svgHeight = svgDoc.height;
      svgWidth = svgDoc.width;
    } else {
      svgHeight = svgDoc.viewBox.replace(/^\d+\s\d+\s(\d+\.?[\d])\s(\d+\.?[\d])/, '$2');
      svgWidth = svgDoc.viewBox.replace(/^\d+\s\d+\s(\d+\.?[\d])\s(\d+\.?[\d])/, '$1');
    }
    
    width = parseFloat(svgWidth);
    height = parseFloat(svgHeight);
    
  } catch (err) {
    return;
  }

  return {
    width,
    height,
    type: 'svg',
    alpha,
    animated,
    interlacing,
    size: buffer.length,
  };
};
