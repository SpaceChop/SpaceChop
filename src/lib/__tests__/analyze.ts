import { createReadStream } from 'fs';
import path from 'path';
import toMatchImageSnapshot from '../../test/utils/toMatchImageSnapshot';
import { ImageDefinition, ImageFaceBox, DefinitionRequirement } from '../../types';
import { Format } from '../../types/Format';
import analyze from '../analyze';
import DefinitionRequirement from '../../types/DefinitionRequirement';

expect.extend({ toMatchImageSnapshot });

describe('Analyze', () => {
  describe('Basic files', () => {
    const assets = '../../test/assets';
    const sources: Array<{
      source: string;
      alpha: boolean;
      interlacing: boolean;
      root: string,
      type: Format,
      width: number,
      height: number
      animated: boolean,
      lossy?: boolean,
      faces?: ImageFaceBox[],
      exif?: Partial<{
        Orientation: number,
      }>,
    }> = [{
      source: 'grid.jpg',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'jpeg',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid-no-exif.jpg',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'jpeg',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid-interlaced.jpg',
      alpha: false,
      interlacing: true,
      root: assets,
      type: 'jpeg',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid.png',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'png',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'cat.png',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'png',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'cat-interlaced.png',
      alpha: true,
      interlacing: true,
      root: assets,
      type: 'png',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'cat-no-alpha.png',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'png',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid.gif',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'gif',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid.webp',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'grid-lossy.webp',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'rose-alpha.webp',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 400,
      height: 301,
      animated: false,
    }, {
      source: 'lossy-vp8.webp',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 550,
      height: 368,
      animated: false,
    }, {
      source: 'grid-interlaced.gif',
      alpha: false,
      interlacing: true,
      root: assets,
      type: 'gif',
      width: 100,
      height: 100,
      animated: false,
    }, {
      source: 'animated.gif',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'gif',
      width: 100,
      height: 100,
      animated: true,
    }, {
      source: 'animated.png',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'png',
      width: 100,
      height: 100,
      animated: true,
    }, {
      source: 'animated.webp',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 320,
      height: 240,
      animated: true,
      lossy: false,
    }, {
      source: 'animated-alpha.webp',
      alpha: true,
      interlacing: false,
      root: assets,
      type: 'webp',
      width: 200,
      height: 200,
      animated: true,
      lossy: false,
    }, {
      source: 'small-face.jpg',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'jpeg',
      width: 100,
      height: 126,
      animated: false,
      lossy: false,
      faces: [{
        x: 27,
        y: 32,
        width: 52,
        height: 52,
      }],
    }, {
      source: 'L3.jpg',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'jpeg',
      width: 100,
      height: 67,
      animated: false,
      lossy: false,
      exif: expect.objectContaining({
        Orientation: 3,
      }),
    }, {
      source: 'L7.jpg',
      alpha: false,
      interlacing: false,
      root: assets,
      type: 'jpeg',
      width: 67,
      height: 100,
      animated: false,
      lossy: false,
      exif: expect.objectContaining({
        Orientation: 7,
      }),
    }];

    // create a test for all file types
    for (const {
      source,
      width,
      height,
      alpha,
      interlacing,
      animated,
      type,
      root,
      faces,
      exif,
    } of sources) {
      describe(`analyze ImageDefinition for ${source}`, () => {
        const expected: ImageDefinition = {
          width,
          height,
          type,
          alpha,
          interlacing,
          animated,
          ...faces && { faces },
          ...exif && { exif },
        };
        const stream = createReadStream(path.join(__dirname, root, source));

        it('should return valid ImageDefinition', async () => {
          const requirements: { [key: number]: DefinitionRequirement } = {
            ...faces && { 0: 'faces' },
            ...exif && { 1: 'exif' },
          };

          const recieved = await analyze(stream, requirements);
          expect(recieved).toEqual(
            expect.objectContaining(expected),
          );
        });
      });
    }
  });

  describe('Bad files', () => {
    const assets = '../../test/assets';
    const sources: Array<{
      source: string;
      root: string,
    }> = [{
      source: 'zero-byte-file',
      root: assets,
    }];

    // create a test for all file types
    for (const {
      source,
      root,
    } of sources) {
      describe(`analyze ImageDefinition for ${source}`, () => {
        const stream = createReadStream(path.join(__dirname, root, source));

        it('should reject', async () => {
          await expect(analyze(stream, [])).rejects.toMatchSnapshot();
        });
      });
    }
  });
});
