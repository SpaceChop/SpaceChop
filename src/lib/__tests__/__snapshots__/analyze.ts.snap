// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyze Bad files analyze ImageDefinition for corrupted.png should reject 1`] = `"Unsupported type"`;

exports[`Analyze Bad files analyze ImageDefinition for text-file.txt should reject 1`] = `"Unsupported type"`;

exports[`Analyze Bad files analyze ImageDefinition for zero-byte-file should reject 1`] = `"Unsupported type"`;
